include/engine/routing_algorithms/routing_base.hpp

```diff
@@ -524,7 +528,11 @@ template <class DataFacadeT, class Derived> class BasicRoutingInterface
                                           std::vector<NodeID> &packed_path) const
     {
         NodeID current_node_id = middle_node_id;
-        while (current_node_id != search_heap.GetData(current_node_id).parent)
+        // all initial nodes will have itself as parent
+        // all incomplete paths (e.g. paths ending at the core boundary) will have SPECIAL_NODEID
+        // as parent
+        while (current_node_id != search_heap.GetData(current_node_id).parent &&
+               search_heap.WasInserted(search_heap.GetData(current_node_id).parent))
```

Comment:
Than we still have a problem. When we have two core nodes for the start/end, we still need to find loops that have been contracted. Or can't this ever happen? I guess it is a rare case, but the fix right now introduces a potential problem here.

Loops should only be searched for on `distance < 0` or if forced. So it looks like we might force loops at the wrong location?
Or could this `distance < 0` result from via-routes?

I guess since the map-matchings are the bigger problem over the loops on core nodes, we can merge this. You should probably add some documentation in the code to why this is important to do as it is and open an issue for core-node loops referencing these changes.